src/Microsoft.AspNetCore.Razor.VSCode/src/RazorDocumentSynchronizer.ts

```diff
@@ -24,164 +25,257 @@ export class RazorDocumentSynchronizer {
 
-            if (context && csharpDocument.hostDocumentSyncVersion >= context.documentVersion) {
+        if (!synchronizationContexts) {
+            return;
+        }
+
+        for (const context of synchronizationContexts) {
+            if (context.projectedDocument.projectedDocumentSyncVersion === projectedTextDocumentVersion) {
                 if (this.logger.verboseEnabled) {
-                    this.logger.logVerbose(
-                        `${context.logIdentifier} - Notify Success: CSharp updated to ${context.documentVersion}.`);
+                    const li = context.logIdentifier;
+                    const ptdv = projectedTextDocumentVersion;
+                    this.logger.logVerbose(`${li} - Projected text document synchronized to ${ptdv}.`);
                 }
-                context.synchronized(true);
+                context.projectedTextDocumentSynchronized();
             }
-        } else if (event.kind === RazorDocumentChangeKind.closed) {
-            const uriPath = getUriPath(event.document.uri);
-            const context = this.synchronizations[uriPath];
+        }
+    }
 
-            if (context) {
-                if (this.logger.verboseEnabled) {
-                    this.logger.logVerbose(
-                        `${context.logIdentifier} - Notify Failed: Document closed.`);
-                }
+    private documentChanged(event: IRazorDocumentChangeEvent) {
+        let projectedDocument: IProjectedDocument;
+        if (event.kind === RazorDocumentChangeKind.csharpChanged) {
+            projectedDocument = event.document.csharpDocument;
+        } else if (event.kind === RazorDocumentChangeKind.htmlChanged) {
+            projectedDocument = event.document.htmlDocument;
+        } else {
+            return;
+        }
+
+        const hostDocumentSyncVersion = projectedDocument.hostDocumentSyncVersion;
+        if (hostDocumentSyncVersion === null) {
+            return;
+        }
 
-                context.synchronized(false);
+        const documentKey = getUriPath(projectedDocument.uri);
+        const synchronizationContexts = this.synchronizations[documentKey];
+        if (!synchronizationContexts) {
+            return;
+        }
+
+        for (const context of synchronizationContexts) {
+            if (context.toHostDocumentVersion === projectedDocument.hostDocumentSyncVersion) {
+                context.projectedDocumentSynchronized();
             }
+        }
+    }
+
+    private getProjectedTextDocumentVersion(textDocument: vscode.TextDocument) {
+        // Logic defined in this method is heavily dependent on the functionality in the projected
+        // document content providers to append versions to the end of text documents.
 
+        if (textDocument.lineCount <= 0) {
+            return null;
         }
+
+        const lastLine = textDocument.lineAt(textDocument.lineCount - 1);
+
+        if (!lastLine.text.startsWith('// ')) {
```

Comment:
Not sure I buy this logic.

1. If there's a possibility that the last line might *not* start with `//`, then that implies the system doesn't always insert the `// (number)` line.
2. As a consequence, it's possible that the last line was something the developer wrote themselves.
3. The thing the developer wrote themselves might start with `// ` but not be followed by a number.
4. In this case, we'll get a `NaN` on line 264, or possibly even a random other number. Not sure how to reason about what the effect of this would be.

If we have to distinguish between system-added lines and developer-written ones, could we at least make accidental conflicts unlikely by changing the pattern to something like `// [VS Code Razor sync version {number}]`?

Reply:
> 1. If there's a possibility that the last line might _not_ start with `//`, then that implies the system doesn't always insert the `// (number)` line.

Not possible because we own the document generation. This is only looking at virtual documents that we serve (and ensure that the last line is `// (number)`



> 2\. As a consequence, it's possible that the last line was something the developer wrote themselves.

Can't because we append a line after the last line that the developer wrote.

> 3\. The thing the developer wrote themselves might start with `// ` but not be followed by a number.

That's fine, it wont be the last line.

> 4\. In this case, we'll get a `NaN` on line 264, or possibly even a random other number. Not sure how to reason about what the effect of this would be.

Because of the above (don't need to distinguish) we don't need to worry about this.

Reply:
So you don't need line 259 then?

Reply:
Fair, yup don't really need that 😄